dao/src/main/java/com/epam/esm/dao/dao/impl/GiftCertificateDaoImpl.java

```diff
@@ -1,41 +1,144 @@
+            " certificate.duration = ?, certificate.create_date = ?, "
+            + "last_update_date = ? WHERE certificate.certificate_id = ?";
+    private static final String FIND_BY_NAME = "SELECT certificate.certificate_id,certificate.certificate_name, " +
+            "certificate.description, certificate.price,certificate.duration,certificate.create_date, " +
+            "certificate.last_update_date, " +
+            " tag.id_tag,tag.name_tag" +
+            " FROM certificate " +
+            " LEFT JOIN certificate_tag ON certificate.certificate_id=certificate_tag.certificate_id " +
+            " LEFT JOIN tag ON certificate_tag.id_tag=tag.id_tag " +
+            " WHERE certificate.certificate_name=?";
 
     @Autowired
-    public GiftCertificateDaoImpl(JdbcTemplate jdbcTemplate) {
+    public GiftCertificateDaoImpl(JdbcTemplate jdbcTemplate, CertificateMapper certificateMapper, CertificateExtractor certificateExtractor) {
         this.jdbcTemplate = jdbcTemplate;
+        this.certificateMapper = certificateMapper;
+        this.certificateExtractor = certificateExtractor;
     }
 
     @Override
     public Optional<GiftCertificate> findEntityById(long id) {
-        throw new UnsupportedOperationException("Not supported yet.");
+        List<GiftCertificate> certificateList = jdbcTemplate.query(FIND_CERTIFICATE_BY_ID, certificateExtractor, id);
+        assert certificateList != null;
+        return certificateList.isEmpty() ? Optional.empty() : Optional.of(certificateList.get(0));
     }
 
     @Override
     public boolean delete(long id) {
-        throw new UnsupportedOperationException("Not supported yet.");
+        return jdbcTemplate.update(DELETE_BY_ID, id) > 0;
     }
 
     @Override
-    public int create(GiftCertificate entity) {
-        throw new UnsupportedOperationException("Not supported yet.");
+    public boolean deleteLinkTagById(long id) {
+        return jdbcTemplate.update(DELETE_LINK_TAG_BY_ID, id) > 0;
     }
 
     @Override
-    public int update(GiftCertificate entity) {
-        throw new UnsupportedOperationException("Not supported yet.");
+    public boolean addLinkTag(long certificateId, long tagId) {
+        PreparedStatementCreator preparedStatementCreator = connection -> {
+            PreparedStatement preparedStatement = connection.prepareStatement(ADD_LINK_ONE_TAG);
+            preparedStatement.setLong(1, certificateId);
+            preparedStatement.setLong(2, tagId);
+            return preparedStatement;
+        };
+        return jdbcTemplate.update(preparedStatementCreator) > 0;
+    }
+
+    @Override
+    public List<GiftCertificate> findEntityByParams(Map<String, String> groupParams) {
+        String sql = FIND_CERTIFICATES_BY_PARAMS +
+                BuilderSql.buildFindAndSortCertificateByParameter(groupParams);
+        return jdbcTemplate.query(sql, certificateExtractor);
+
```

Comment:
there should be an empty line after return statements (please follow Java Code Convention)

Reply:
done